### ISSUE TITLE
(New feature request)If IP is in blacklist, delete this response then use next DNS instead of return NXDOMAIN.

### ISSUE THREAD

**Issue by iamsurfing** [folbricht/routedns#22]
If IP is in blacklist, delete this response then use next DNS instead of return NXDOMAIN.

**Comment by folbricht** [comment]
Can you provide more detail on this? What does "use next DNS" mean in this context? It sounds a little like you're looking for something similar to #21 and/or #10.

**Comment by cbuijs** [comment]
My take on this and how I read it:

When you resolve something, you get back an RRSet of 3 IP-Addresses, one of those is blocklisted, and instead of blocking the whole request, you strip out the offending IP and return the remaining IP's in the response. (Block the whole request when all IP's offend).

It is a good case to block particular IP-Ranges, I use it (with DeugNietS) to make sure my clients do not connect to some questionable countries/ranges but still have services running that we need. Mostly IoT, Camera's, some apps and the likes. Protect the young ones ;-).

**Comment by folbricht** [comment]
Ah, thanks for the explanation. So this would be for the response blocklist to filter out the matching RRs rather than blocking the whole request. That makes sense.

**Comment by cbuijs** [comment]
Yes, that is how interpreted it.

Cool feature anyway :-).

**Comment by folbricht** [comment]
That's been implemented on the issue-22 branch. There's now a boolean `filter` option on it that changes it from blocking into filtering mode.

```toml
[groups.cloudflare-blocklist]
type             = "response-blocklist-name"
resolvers        = ["cloudflare-dot"]
blocklist-format = "domain"
blocklist        = [
  'ns.evil.com.',
  '*.acme.test',
]
filter = true # Set to true if the response RRs should be filtered rather than returning NXDOMAIN (default)
```

**Comment by cbuijs** [comment]
It works, but.... When you have cname chain (try to resolve `api.allekabels.nl`), it breaks the chain and provide invalid/incomplete answers.

It might be more logical to keep it to cidr and only for A/AAAA type records and non chained records (CNAME is probably the only one actually that will break).

Also for the `name` and `cidr` blocklist types, a `blocklist-resolver` feature like blocklist-v2 would be nice. Also a NXDOMAIN when no records left after a filtering.

**Comment by iamsurfing** [comment]
I am sorry my computer was repaired, so I can not reply you at once.
What I say is not about RRs.
![draw](https://user-images.githubusercontent.com/16305191/82143169-649f2100-9874-11ea-89f3-904662788b47.png)
I think this draw can express my opinion.

**Comment by folbricht** [comment]
Yes, makes sense, I'll remove it from the `-name` blocklist type and make it NXDOMAIN if no answer is left after filtering.

@iamsurfing If I understand it correctly, you also would like the `blocklist-resolver` feature mentioned above. Basically
- Send the query to the default resolver
- If there's a match in the response, discard the whole answer and send the query again, but to the `blocklist-resolver` (if defined).

**Comment by folbricht** [comment]
Removed `filter` support from response-blocklist-name to avoid the issues above. Also, both response blocklists now support `blocklist-resolver` which works like this:

```toml
[resolvers.default-dot]
address = "1.1.1.1:853"
protocol = "dot"

[resolvers.alternate-dot]
address = "8.8.8.8:853"
protocol = "dot"

[groups.cloudflare-blocklist]
type               = "response-blocklist-cidr"
resolvers          = ["default-dot"] # Default resolver, all queries are sent here first
blocklist-resolver = "alternate-dot" # Use this resolver when a response matches the blocklist
blocklist          = [
  '127.0.0.0/24',
  '157.240.0.0/16',
]
```
Note that things sent to blocklist-resolver after a match will not be re-evaluated again against the blocklist.

@iamsurfing Does the above solve your use-case?

**Comment by iamsurfing** [comment]
This configuration is what I want, so I have rebuilt and tested it.
However, blocklist-resolver seems not work. If match IP, it will still return NXDOMAIN.

**Comment by folbricht** [comment]
This is still sitting in a PR https://github.com/folbricht/routedns/pull/26 for now, did you build from the `issue-22` branch?

**Comment by iamsurfing** [comment]
It is perfect. Thank you.

**Comment by folbricht** [comment]
Merged in #26
